### PR TITLE
Improve the way settings are managed

### PR DESCRIPTION
**Pull request by sdispater**
## Pull Request Check List

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

This PR is a refactoring and improvement of the configuration part of Poetry which introduces the following changes.

### Local, per-project settings

The `config` command now supports a `--local` option which will store the given settings to a local `poetry.toml` file.

```bash
poetry config --local virtualenvs.in-project true
```

### Configuration via environment variables

Poetry can now be configured using environment variables. This is particularly useful in CI environments and avoids having to create configuration files.

The environment variables must be prefixed by `POETRY_` and are comprised of the uppercase
name of the setting with dots and dashes replaced by underscore, here is an example:

```bash
export POETRY_VIRTUALENVS_PATH=/path/to/virtualenvs/directory
```

This also works for secret settings, like credentials:

```bash
export POETRY_HTTP_BASIC_MY_REPOSITORY_PASSWORD=secret
```

### The `settings.` prefix is no longer necessary

This one is a breaking change. The settings are now stored differently and no longer needs to be prefixed by `settings.`. If you have already configured settings you will need to configure them again.

### CODE REVIEW

File: poetry/config/config.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from typing import Optional
+
+from poetry.locations import CACHE_DIR
+from poetry.utils._compat import Path
+from poetry.utils._compat import basestring
+
+_NOT_SET = object()
+
+boolean_validator = lambda val: val in {"true", "false", "1", "0"}
+boolean_normalizer = lambda val: True if val in ["true", "1"] else False
```
Comment (yedpodtrzitko) [inline poetry/config/config.py]:
```suggestion
boolean_normalizer = lambda val: val in {"true", "1"}
```

Comment (sdispater) [reply poetry/config/config.py]: Fixed. Thanks!

**Comment by levrik** [comment]
I'm quite surprised that the local configuration is written to its own file instead of using `pyproject.toml`. What's the reason for that?

**Comment by notpushkin** [comment]
Sorry for bumping this but I'm interested in this decision, too. I understand it allows storing per-project settings while git-ignoring the `poetry.toml` file, but for some projects, storing .venv relatively to the repository is a must (e. g. some scripts depend on this, as noted in https://github.com/python-poetry/poetry/issues/618). Is there a way to support reading from both `pyproject.toml` and `poetry.toml`?

**Comment by austinmw** [comment]
I've read through the various related issues, but is it still not possible to specify a local venv from the pyproject.toml file? For example:

```toml
[settings.virtualenvs]
in-project = "true"
```

**Pull request merged**